infrastructure/crypto/src/ristretto/ristretto_keys.rs

```diff
@@ -0,0 +1,347 @@
+/// use crypto::keys::SecretKeyFactory;
+/// use crypto::common::ByteArray;
+/// use rand;
+///
+/// let mut rng = rand::OsRng::new().unwrap();
+/// let _k1 = SecretKey::from_bytes(&[1,0,0,0, 0,0,0,0, 0,0,0,0, 0,0,0,0, 0,0,0,0, 0,0,0,0, 0,0,0,0, 0,0,0,0]);
+/// let _k2 = SecretKey::from_hex(&"100000002000000030000000040000000");
+/// let _k3 = SecretKey::random(&mut rng);
+/// ```
+#[derive(PartialEq, Eq, Clone, Copy, Debug)]
+pub struct RistrettoSecretKey(pub(crate) Scalar);
+
+const SCALAR_LENGTH: usize = 32;
+const PUBLIC_KEY_LENGTH: usize = 32;
+
+impl SecretKey for RistrettoSecretKey {
+    fn key_length() -> usize {
+        SCALAR_LENGTH
+    }
+}
+
+impl ByteArray for RistrettoSecretKey {
+    /// Create a secret key on the Ristretto255 curve using the given little-endian byte array. If the byte array is
+    /// not exactly 32 bytes long, `from_bytes` returns an error. This function is guaranteed to return a valid key
+    /// in the group since it performs a mod _l_ on the input.
+    fn from_bytes(bytes: &[u8]) -> Result<RistrettoSecretKey, ByteArrayError>
+    where Self: Sized {
+        if bytes.len() != 32 {
+            return Err(ByteArrayError::IncorrectLength);
+        }
+        let mut a = [0u8; 32];
+        a.copy_from_slice(bytes);
+        let k = Scalar::from_bytes_mod_order(a);
+        Ok(RistrettoSecretKey(k))
+    }
+
+    /// Return the byte array for the secret key in little-endian order
+    fn to_bytes(&self) -> &[u8] {
+        self.0.as_bytes()
+    }
+}
+
+/// Return a random secret key on the `ristretto255` curve using the supplied CSPRNG.
+impl SecretKeyFactory for RistrettoSecretKey {
+    fn random<R: CryptoRng + Rng>(rng: &mut R) -> Self {
+        RistrettoSecretKey(Scalar::random(rng))
+    }
+}
+
+/// The [PublicKey](trait.PublicKey.html) implementation for `ristretto255` is a thin wrapper around the dalek
+/// library's [RistrettoPoint](struct.RistrettoPoint.html).
+///
+/// ## Creating public keys
+/// Both [PublicKey](trait.PublicKey.html) and [ByteArray](trait.ByteArray.html) are implemented on
+/// `RistrettoPublicKey` so all of the following will work:
+/// ```edition2018
+/// use crypto::ristretto::{ RistrettoPublicKey, RistrettoSecretKey as SecretKey };
+/// use crypto::common::ByteArray;
+/// use crypto::keys::{ PublicKey, SecretKeyFactory };
+/// use rand;
+///
+/// let mut rng = rand::OsRng::new().unwrap();
+/// let _p1 = RistrettoPublicKey::from_bytes(&[224, 196, 24, 247, 200, 217, 196, 205, 215, 57, 91, 147, 234, 18, 79, 58, 217,
+/// 144, 33, 187, 104, 29, 252, 51, 2, 169, 217, 154, 46, 83, 230, 78]);
+/// let _p2 = RistrettoPublicKey::from_hex(&"e882b131016b52c1d3337080187cf768423efccbb517bb495ab812c4160ff44e");
+/// let sk = SecretKey::random(&mut rng);
+/// let _p3 = RistrettoPublicKey::from_secret_key(&sk);
+/// ```
+#[derive(Clone, Copy, Debug)]
+pub struct RistrettoPublicKey {
+    pub(crate) point: RistrettoPoint,
+    pub(crate) compressed: CompressedRistretto,
```

Comment:
I'd love some opinions on the approach I took here. It's got a bit of code smell tbh, since keeping a copy of the compressed and expanded point is obviously redundant; but keeping only one would require some costly CPU cycles to convert every time you wanted to do a calculation (requires uncompressed) or serialisation/communication (requires compressed).

As it stands the two versions should never go out of sync because the values shouldn't be exposed, but still.

Reply:
I am pro keeping both. But I agree on the code smell point. 
I would remove the pub access of both, and only access both with getters and setters. 
That way you force the two to stay in sync.